### ISSUE TITLE
Proposal: Versioning Workflow

### ISSUE THREAD

**Issue by ethernetdan** [redspread/spread#122]
We would like to introduce a workflow for distributed versioning of Kubernetes objects. Initially the workflow and its interface will mirror Git (branching, committing, pushing, etc) but we should improve on its UX where we can. I definitely would like to see the `checkout` behemoth broken up into multiple commands.

To kick off brainstorming here are my thoughts on what this workflow could look like.

### Example
**Initialize spread repository**
```
$ spread init
Created empty Spread repository in /Users/dan/example/.spread/
```
**Create objects on cluster**
```
$ kubectl create -f "rc.yml"
replicationcontroller "example" created
$ kubectl create -f "service.yml"
service "example" created
```

**Stage objects**
```
$ spread add rc/example
$ spread add service/example
```
Objects are exported from the cluster, serialized to binary, and staged to be committed.

**Confirm**
```
$ spread status
On branch master
Changes to be committed:
  (use "spread reset HEAD <object name>..." to unstage)
        new rc: default/example
        new service:   default/example
```
Check that the correct changes are being made. Similarly `spread diff HEAD` would show the fields being created.

**Commit**
```
$ spread commit -m "added example replication controller"
[master (root-commit) da5a6d7] commit
 2 objects changed, 45 fields added
 create default/rc/example
 create default/service/example
```

**Setup production remote**
```
$ spread remote add prod spread@redspread.com:ethernetdan/example.spread
```
Let's assume that the public key for the host that is being developed on have already been configured on the remote to allow access.

**Push to production cluster**
```
$ spread push prod master
Counting objects: 3, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (2/2), done.
Writing objects: 100% (3/3), 262 bytes | 0 bytes/s, done.
Total 3 (delta 1), reused 0 (delta 0)
To spread@redspread.com:ethernetdan/example.spread
 * [new branch]      master -> master
```
We probably want to use different terminology here to distinguish between Git and Kubernetes objects. 

**Deploy to cluster**
(While SSH'd on server that was pushed to)
```
$ cd ~/ethernetdan/example.spread
$ spread deploy master
Deploying 2 objects using the default kubectl context.
Waiting for load balancer deployment...
Service 'default/service' available at:  104.154.49.200
Deployment successful!
```
We probably want to include an easy way to deploy from remote machines (not requiring the user to SSH themselves to deploy)

**Comment by ethernetdan** [comment]
Additionally, files should be able to be staged like:
```
spread add -f "rc.yml"
spread add -f "service.yml"
```

**Comment by mfburnett** [comment]
Let's also include an add all function, something like:
$ spread add -f -A

**Comment by ethernetdan** [comment]
We need to figure out how namespaces figure in, a solution that could work is borrow from kubectl and have a `--namespace=` flag

**Comment by colhom** [comment]
@ethernetdan  Namespaces are functional components in the sense that things like resource limits and ABAC policies can be tied to them.  I've found that the namespaces of a cluster (and which resources they contain) should be versioned / auditable along as well.
 
Maybe `spread` could detect `namespace` yaml files, and all resources could declare which namespace they belong to in `metadata`. Then a `spread deploy` can take care of an entire cluster, if desired.

**Comment by ethernetdan** [comment]
@colhom really good point here, I was working off the assumption that the namespace "default" would be implicitly included but I'm convinced that it should itself be versioned.

Should we not allow objects to be staged if the namespace they specify hasn't been staged yet?

**Comment by mfburnett** [comment]
Implemented with #165
